### PR TITLE
Update Vulkan README.md and add support for vkResetDescriptorPool

### PR DESCRIPTION
**Pull request by Qining**
Added forward declaration for internal classes, so that on CPP side we won't have undefined-type error.

### CODE REVIEW

File: gapis/gfxapi/templates/api_types.h.tmpl [hunk 1]
```diff
@@ -64,6 +64,9 @@
   {{range $p := $.Pseudonyms}}
     {{Template "DeclareType" $p}}
   {{end}}
+  {{range $c := $.Classes}}
+    {{if GetAnnotation $c "internal"}} class {{$c.Name}}; {{end}}
+  {{end}}
```
Comment (Qining) [inline gapis/gfxapi/templates/api_types.h.tmpl]: @ben-clayton @AWoloszyn Not sure if this is acceptable, but this does solve the undefined-type problem when compiling CPP

File: gapis/gfxapi/vulkan/vulkan.api [hunk 2]
```diff
@@ -9132,6 +9142,7 @@ enum RecordingState {
   @unused VkDescriptorPoolCreateFlags     Flags
   @unused u32                             MaxSets
   @unused map!(u32, VkDescriptorPoolSize) Sizes
+  @unused map!(VkDescriptorSet, ref!DescriptorSetObject) DescriptorSets
```
Comment (AWoloszyn) [inline gapis/gfxapi/vulkan/vulkan.api]: If you just make this a map of u32->VkDescriptorSet you will avoid the circular dependency. We probably don't need the refs to the Objects here.

Comment (Qining) [inline gapis/gfxapi/vulkan/vulkan.api]: Right, but this means we will need to traverse through all the descriptorsets in a pool when we want to free a specific descriptorset.

**Comment by Qining** [comment]
@AWoloszyn Updated.

DescriptorPoolObject now has a map from VkDescriptorSet -> Ref!DescriptorSetObject
DescriptorSetObject now does not have Ref!DescriptorPoolObject, but a VkDescriptorPool.

File: gapis/gfxapi/vulkan/vulkan.api [hunk 3]
```diff
@@ -4607,6 +4607,11 @@ cmd VkResult vkResetDescriptorPool(
     VkDevice                   device,
     VkDescriptorPool           descriptorPool,
     VkDescriptorPoolResetFlags flags) {
+  sets := DescriptorPools[descriptorPool].DescriptorSets
```
Comment (Qining) [inline gapis/gfxapi/vulkan/vulkan.api]: If I understand it correctly, this should create a copy, so we won't erasing items while iterating them.

Comment (AWoloszyn) [inline gapis/gfxapi/vulkan/vulkan.api]: I think so, although I would look at the generated golang code.
I have a feeling that in golang we don't actually copy the map, but reference it.

Comment (Qining) [inline gapis/gfxapi/vulkan/vulkan.api]: Interesting.
In CPP side, it is a copy, so no problem.
In Golang side, it is not a copy, however, it is **safe** to remove item while iterating!
https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-golang-map-within-a-range-loop

Comment (ben-clayton) [inline gapis/gfxapi/vulkan/vulkan.api]: Can you make a note of the behaviors you're relying on?
If I ever get around to finishing the LLVM compiler, I'm going to have to be careful to either correct this code or keep the same behaviors.

Comment (Qining) [inline gapis/gfxapi/vulkan/vulkan.api]: OK, I will add comments there and rebase on top of current master.

**Comment by Qining** [comment]
@ben-clayton @AWoloszyn Rebase DONE and updated. All comments resolved.

**Pull request merged**